registry/storage/driver/swift/swift.go

```diff
@@ -0,0 +1,547 @@
+	if offset >= currentLength {
+		for offset-currentLength >= chunkSize {
+			// Insert a block a zero
+			d.Conn.ObjectPut(d.Container, getSegment(),
+				bytes.NewReader(zeroBuf), false, "",
+				d.getContentType(), nil)
+			currentLength += chunkSize
+			partNumber++
+		}
+
+		cursor = currentLength
+		paddingReader = bytes.NewReader(zeroBuf)
+	} else if offset-cursor > 0 {
+		// Offset is inside the current segment : we need to read the
+		// data from the beginning of the segment to offset
+		file, _, err := d.Conn.ObjectOpen(d.Container, getSegment(), false, nil)
+		defer file.Close()
+		paddingReader = file
+
+		if err != nil {
+			return 0, parseError(getSegment(), err)
+		}
+	}
+
+	readers := []io.Reader{}
+	if paddingReader != nil {
+		readers = append(readers, io.LimitReader(paddingReader, offset-cursor))
+	}
+	readers = append(readers, io.LimitReader(reader, chunkSize-(offset-cursor)))
+	multi = io.MultiReader(readers...)
+
+	writeSegment := func(segment string) (finished bool, bytesRead int64, err error) {
+		currentSegment, err := d.Conn.ObjectCreate(d.Container, segment, false, "", d.getContentType(), nil)
+		if err != nil {
+			return false, bytesRead, parseError(path, err)
+		}
+
+		n, err := io.Copy(currentSegment, multi)
+		if err != nil {
+			return false, bytesRead, parseError(path, err)
+		}
+
+		if n > 0 {
+			defer currentSegment.Close()
+			bytesRead += n - max(0, offset-cursor)
+		}
+
+		if n < chunkSize {
+			// We wrote all the data
+			if cursor+n < currentLength {
+				// Copy the end of the chunk
+				headers := make(swift.Headers)
+				headers["Range"] = "bytes=" + strconv.FormatInt(cursor+n, 10) + "-" + strconv.FormatInt(cursor+chunkSize, 10)
+				file, _, err := d.Conn.ObjectOpen(d.Container, d.swiftPath(path), false, headers)
+				if err != nil {
+					return false, bytesRead, parseError(path, err)
+				}
+
+				_, copyErr := io.Copy(currentSegment, file)
+
+				if err := file.Close(); err != nil {
+					return false, bytesRead, parseError(path, err)
+				}
+
+				if copyErr != nil {
+					return false, bytesRead, parseError(path, copyErr)
+				}
+			}
+
+			return true, bytesRead, nil
+		}
+
+		multi = io.LimitReader(reader, chunkSize)
+		cursor += chunkSize
+		partNumber++
+
+		return false, bytesRead, nil
+	}
+
+	finished := false
+	read := int64(0)
+	bytesRead := int64(0)
+	for finished == false {
+		finished, read, err = writeSegment(getSegment())
+		bytesRead += read
+		if err != nil {
+			return bytesRead, err
+		}
+	}
+
+	return bytesRead, nil
+}
+
+// Stat retrieves the FileInfo for the given path, including the current size
+// in bytes and the creation time.
+func (d *driver) Stat(ctx context.Context, path string) (storagedriver.FileInfo, error) {
+	info, _, err := d.Conn.Object(d.Container, d.swiftPath(path))
+	if err != nil {
+		return nil, parseError(path, err)
+	}
+
+	fi := storagedriver.FileInfoFields{
+		Path:    path,
+		IsDir:   info.ContentType == directoryMimeType,
```

Comment:
Are the object mimetypes sufficiently protected to ensure that users cannot set to this value and corrupt or change the filesystem? For example, what if we started writing mimetypes to these objects from the higher level application?